File: python/Sorting/selectionSort_v2.py

def selection_sort(arr):
    n = len(arr)
    if n<=1: return arr

    for i in range(n-1):
        mn = i # at first assume that i'th index has the min element
        for j in range(i+1,n):
            if arr[j]<arr[mn]: mn = j # find if there is any element less than ours assumed one
        if i!=mn: arr[mn],arr[i] = arr[i],arr[mn] # if found than swap the elements

arr = [1,4,3,2] # unsorted array
selection_sort(arr) # sort the array in itself
print(arr)

File: cpp/Sorting/selectionSort_v2.cpp
#include <iostream>
#include <vector>
using namespace std;

#define vi vector<int>
#define vb vector<bool>

void selection_sort(vi &arr){
    int n = arr.size();
    if(n<=1) return;
    
    for(int i=0; i<n-1; i++){
        int mn = i; // at first assume that i'th index has the min element
        for(int j=i+1; j<n; j++) if(arr[j]<arr[mn]) mn = j; // find if there is any element less than ours assumed one
        if(i!=mn) swap(arr[i],arr[mn]); // if found than swap the elements
    }
}

int main(){
    freopen("input.txt","r",stdin);
    int n;
    cin >> n;
    vi arr(n);
    for(int i=0;i<n;i++) cin >> arr[i];
    selection_sort(arr); // sort the array in itself
    for(int i:arr) cout << i << ' ';
    cout << '\n';
}